newrelic/hooks/mlmodel_langchain.py

```diff
@@ -0,0 +1,174 @@
+    "LLMRails",
+    "Marqo",
+    "MatchingEngine",
+    "Meilisearch",
+    "Milvus",
+    "MomentoVectorIndex",
+    "MongoDBAtlasVectorSearch",
+    "MyScale",
+    "MyScaleSettings",
+    "Neo4jVector",
+    "OpenSearchVectorSearch",
+    "PGEmbedding",
+    "PGVector",
+    "Pinecone",
+    "Qdrant",
+    "Redis",
+    "Rockset",
+    "SKLearnVectorStore",
+    "ScaNN",
+    "SemaDB",
+    "SingleStoreDB",
+    "SQLiteVSS",
+    "StarRocks",
+    "SupabaseVectorStore",
+    "Tair",
+    "TileDB",
+    "Tigris",
+    "TimescaleVector",
+    "Typesense",
+    "USearch",
+    "Vald",
+    "Vearch",
+    "Vectara",
+    "VespaStore",
+    "Weaviate",
+    "Yellowbrick",
+    "ZepVectorStore",
+    "Zilliz",
+    "TencentVectorDB",
+    "AzureCosmosDBVectorSearch",
+    "VectorStore",
+]
+
+
+def bind_similarity_search(query, k, *args, **kwargs):
+    return query, k
+
+
+def wrap_similarity_search(wrapped, instance, args, kwargs):
+    transaction = current_transaction()
+    if not transaction:
+        return wrapped(*args, **kwargs)
+
+    request_query, request_k = bind_similarity_search(*args, **kwargs)
+    with DatastoreTrace(
+        product="LangChain_vectorstore", target=request_query, operation="similarity_search", source=wrapped
+    ) as dt:
+        try:
+            response = wrapped(*args, **kwargs)
+        except Exception as exc:
+            # Error logic goes here
+            pass
+
+    if not response:
+        return response
+
+    # LLMVectorSearch
+    available_metadata = get_trace_linking_metadata()
+    span_id = available_metadata.get("span.id", "")
+    trace_id = available_metadata.get("trace.id", "")
+    transaction_id = transaction.guid
+    id = str(uuid.uuid4())
+    request_query, request_k = bind_similarity_search(*args, **kwargs)
+    duration = dt.duration
+    response_number_of_documents = len(response)
+
+    LLMVectorSearch_dict = {
+        "span_id": span_id,
+        "trace_id": trace_id,
+        "transaction_id": transaction_id,
+        "id": id,
+        "vendor": "LangChain",
+        "request.query": request_query,
+        "request.k": request_k,
+        "duration": duration,
+        "response.number_of_documents": response_number_of_documents,
+    }
+
+    transaction.record_custom_event("LlmVectorSearch", LLMVectorSearch_dict)
+
+    # LLMVectorSearchResult
+    for index, doc in enumerate(response):
+        search_id = str(uuid.uuid4())
+        sequence = index
+        page_content = getattr(response[0], "page_content", "")
+        metadata = getattr(response[0], "metadata", "")
+
+        metadata_dict = {}
+        for key, value in metadata.items():
+            new_key = "metadata.%s" % key
+            metadata_dict[new_key] = value
+
+        LLMVectorSearchResult_dict = {
+            "search_id": search_id,
+            "sequence": sequence,
+            "page_content": page_content,
+        }
+
+        LLMVectorSearchResult_dict.update(LLMVectorSearch_dict)
```

Comment:
I've updated this section of the CDD to not repeat all the data on each SearchResult event. I've also added an appName and ingest_source to each of these events. Sorry to change it on you. Your PR made me notice that it was repeating all the data on the SearchResult event which seemed wasteful and incorrect.